flashinfer/gemm.py

```diff
@@ -1942,70 +2093,128 @@ def mm_fp4(
-            _torch_data_type_to_cudnn_data_type(out_dtype),
-            block_size,
-            a.device,
-            alpha is not None,
-            use_nvfp4,
-        )
+    # Auto-select the best backend
+    if backend == "auto":
+        cuda_major, _ = get_cuda_version(a.device)
+        cc_major, cc_minor = get_compute_capability(a.device)
+        # If cuda version is 13 or greater:
+        # cudnn is more performant if cudnn version is 9.14 or greater.
+        if cuda_major >= 13 and cudnn.backend_version() >= 91400:
+            candidate_backends = ("cudnn", "cutlass")
+        # Otherwise, prioritize cutlass
+        else:
+            candidate_backends = ("cutlass", "cudnn")
+
+        # Filter to only supported backends for this compute capability
+        # Note: The requirement function already validated that at least one backend is supported
+        backends = []
+        for candidate in candidate_backends:
+            # mypy requires explicit type casting for the backend literal
+            backend_literal = cast(
+                Literal["cudnn", "trtllm", "cutlass", "auto"], candidate
+            )
+            try:
+                _check_mm_fp4_problem_size(
+                    a,
+                    b,
+                    a_descale,
+                    b_descale,
+                    alpha,
+                    out_dtype,
+                    out,
+                    block_size,
+                    use_8x4_sf_layout,
+                    backend_literal,
+                    use_nvfp4,
+                )
+                backends.append(candidate)
+            except Exception:
+                pass
+    else:
+        backends = [backend]
 
-        # execute the fp4 cudnn graph
-        execute_cudnn_gemm_fp4_graph(
-            graph, a, b, a_descale, b_descale, alpha, out, workspace_buffer
-        )
-    elif backend == "trtllm":
-        get_trtllm_fp4_gemm_module().trtllm_fp4_gemm(
-            a,
-            b.T,
-            a_descale,
-            b_descale.T,
-            alpha,
-            out,
-            use_8x4_sf_layout=use_8x4_sf_layout,
-            workspace_buffer=workspace_buffer,
-        )
-    elif backend == "cutlass":
-        # cutlass require uint8 scale when a/b is fp4 packed uint8.
-        if a.dtype == torch.uint8 and a_descale.dtype == torch.float8_e4m3fn:
-            a_descale = a_descale.view(torch.uint8)
-        if b.dtype == torch.uint8 and b_descale.dtype == torch.float8_e4m3fn:
-            b_descale = b_descale.view(torch.uint8)
-
-        # Dispatch to the correct module based on device architecture
-        major, _ = get_compute_capability(a.device)
-        if major == 12:
-            gemm_module = get_gemm_sm120_module_cutlass_fp4()
+    # At this point, backends contains a supported backend if specified, or all supported backends if backend='auto'.
+    runners = []
+    for cur_backend in backends:
+        if cur_backend == "cudnn":
```

Comment:
Is there a cleaner way than the if-then-else structure? Maybe some dictionary that maps between backend and runner retrieval?